### PR TITLE
tablestore binlog parser udf

### PR DESCRIPTION
**Pull request by okingniko**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(Refer to this [test guide](https://github.com/aliyun/aliyun-emapreduce-sdk/blob/master-2.x/docs/how_to_run_tests.md))

### CODE REVIEW

File: emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+import java.util.UUID
+
+import io.delta.tables.DeltaTable
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.{AnalysisException, Row, SparkSession}
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.streaming.Trigger
+import org.apache.spark.sql.types.{DataTypes, StructField}
+
+object DeltaTableStoreCDC extends Logging {
```
Comment (uncleGen) [inline emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala:29]: move this demo to `examples` module

File: emr-sql/pom.xml [hunk 2]
```diff
@@ -134,6 +134,12 @@
             <artifactId>hadoop-client</artifactId>
             <scope>provided</scope>
         </dependency>
+        <dependency>
```
Comment (uncleGen) [inline emr-sql/pom.xml:137]: ditto, move this dependency to `examples` module

File: emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    _*
+    ) = args
+
+    val checkpointLocation =
+      if (args.length > 7) args(7) else "/tmp/temporary-" + UUID.randomUUID.toString
+
+    val spark =
+      SparkSession.builder
+        .appName("DeltaTableStoreCDC")
+        .master("local[16]")
```
Comment (uncleGen) [inline emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala:56]: better to use `local[*]`

File: emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/FromTableStoreCDCUDF.scala [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.aliyun.udfs.tablestore
+
+import org.apache.spark.internal.Logging
+
+class FromTableStoreCDCUDF extends Logging {
```
Comment (uncleGen) [inline emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/FromTableStoreCDCUDF.scala:22]: useless class?

File: emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/FromTableStoreCDCUDF.scala [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+}
+
+object FromTableStoreCDCUDF {
```
Comment (uncleGen) [inline emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/FromTableStoreCDCUDF.scala:26]: How to use this udf in SQL?

File: emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.aliyun.udfs.tablestore
+
+import java.util.UUID
+
+import io.delta.tables.DeltaTable
```
Comment (uncleGen) [inline emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala:22]: add a line between spark and third-part import

File: emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+
+    val Array(
+    instanceName,
+    tableName,
+    tunnelId,
+    accessKeyId,
+    accessKeySecret,
+    endpoint,
+    maxOffsetsPerChannel,
+    _*
```
Comment (uncleGen) [inline emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala:47]: nit: missing indent

File: emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+      sb.append(" FROM stream")
+      sb.toString
+    }
+
+    val sqlStatement = genQuerySqlString()
+    log.info(s"Current query sql: $sqlStatement")
+
+    val dataStream = spark.sql(
+      sqlStatement
+    )
```
Comment (uncleGen) [inline emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala:147]: one line is better

File: emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/FromTableStoreCDCUDF.scala (hunk 4 above)
Comment (uncleGen) [inline emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/FromTableStoreCDCUDF.scala:22]: the name, i.e. `FromTableStoreCDCUDF`, seems odd, can we use `ResolveTableStoreBinlogUDF`?

File: emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala (hunk 1 above)
Comment (okingniko) [reply emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala:29]: done

Comment (okingniko) [reply emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala:29]: 1

File: emr-sql/pom.xml (hunk 2 above)
Comment (okingniko) [reply emr-sql/pom.xml:137]: 1

File: emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala (hunk 3 above)
Comment (okingniko) [reply emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala:56]: 1

File: emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/FromTableStoreCDCUDF.scala (hunk 5 above)
Comment (okingniko) [reply emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/FromTableStoreCDCUDF.scala:26]: 1

File: emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala (hunk 6 above)
Comment (okingniko) [reply emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala:22]: 1

File: emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala (hunk 7 above)
Comment (okingniko) [reply emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala:47]: 1

File: emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala (hunk 8 above)
Comment (okingniko) [reply emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/DeltaTableStoreCDC.scala:147]: 1

File: emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/FromTableStoreCDCUDF.scala (hunk 4 above)
Comment (okingniko) [reply emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/FromTableStoreCDCUDF.scala:22]: 1

File: emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/ResolveTableStoreBinlogUDF.scala [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+      |       ots_col_parser(col_double, __ots_column_type_col_double) AS col_double,
+      |       ots_col_parser(col_boolean, __ots_column_type_col_boolean) AS col_boolean FROM stream_view;
+      |   PUT 1595990621936075 00008 1595990621 str1 123456 null 3.1415 true
+      |
+    """
+    // scalastyle:on
+
+  // Get actual value according column operation type.
+  def getActualValue[T](origValue: T, columnType: String): T = {
+    columnType match {
```
Comment (uncleGen) [inline emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/ResolveTableStoreBinlogUDF.scala:83]: unified to lower case？

File: examples/src/main/scala/com/aliyun/emr/examples/DeltaTableStoreCDC.scala [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+        DataTypes.createStructField("col5", DataTypes.DoubleType, true),
+        DataTypes.createStructField("col6", DataTypes.BooleanType, true)
+      )
+    )
+    // TotalColumns StructType
+    val totalStructType = DataTypes.createStructType(Array.concat(primaryStructType.fields, attributeStructType.fields))
+
+    // Register hive udf
+    spark.sql("create temporary function ots_col_parser as " +
+      "'org.apache.spark.sql.aliyun.udfs.tablestore.ResolveTableStoreBinlogUDFSuite'")
```
Comment (uncleGen) [inline examples/src/main/scala/com/aliyun/emr/examples/DeltaTableStoreCDC.scala:107]: "ResolveTableStoreBinlogUDFSuite"? typo

File: examples/src/main/scala/com/aliyun/emr/examples/DeltaTableStoreCDC.scala [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+              "col6" -> "source.col6"
+            )
+          )
+          .execute()
+      })
+      .start()
+
+    task.awaitTermination()
+  }
+}
```
Comment (uncleGen) [inline examples/src/main/scala/com/aliyun/emr/examples/DeltaTableStoreCDC.scala:222]: add a new line

File: emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/ResolveTableStoreBinlogUDF.scala (hunk 9 above)
Comment (okingniko) [reply emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/tablestore/ResolveTableStoreBinlogUDF.scala:83]: This is the inner helper method, I think lower camel case style is ok. 
Outside hive method should look like 'xx_xx_xx'.

File: examples/src/main/scala/com/aliyun/emr/examples/DeltaTableStoreCDC.scala (hunk 10 above)
Comment (okingniko) [reply examples/src/main/scala/com/aliyun/emr/examples/DeltaTableStoreCDC.scala:107]: 1

File: examples/src/main/scala/com/aliyun/emr/examples/DeltaTableStoreCDC.scala (hunk 11 above)
Comment (okingniko) [reply examples/src/main/scala/com/aliyun/emr/examples/DeltaTableStoreCDC.scala:222]: 1

**Pull request merged**